File: mobiler/plugins/battery/android/BatteryPlugin.kt
package {{PACKAGE}}

import android.app.Application
import android.content.Context
import android.os.BatteryManager

// A MobilerPlugin (interface declared in Core.kt, same package) that reports the battery
// level as a percentage string ("0".."100"). Installed by `mobiler plugin add battery`.
class BatteryPlugin(private val application: Application) : MobilerPlugin {
    override suspend fun handle(op: String, input: String): PluginResponse {
        if (op != "level") return PluginResponse(false, "unknown op '$op'")
        val bm = application.getSystemService(Context.BATTERY_SERVICE) as? BatteryManager
            ?: return PluginResponse(false, "battery service unavailable")
        val pct = bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY)
        return PluginResponse(true, pct.toString())
    }
}

File: mobiler/plugins/battery/ios/BatteryPlugin.swift
import UIKit

// Reports the battery level as a percentage string ("0".."100"). `PluginResponse` is the
// iOS shell's ABI type. Installed by `mobiler plugin add battery`.
@MainActor
enum BatteryPlugin {
    static func handle(op: String, input: String) async -> PluginResponse {
        guard op == "level" else { return PluginResponse(ok: false, output: "unknown op '\(op)'") }
        UIDevice.current.isBatteryMonitoringEnabled = true
        let level = UIDevice.current.batteryLevel   // 0.0–1.0, or < 0 if unknown
        guard level >= 0 else { return PluginResponse(ok: false, output: "battery level unavailable") }
        return PluginResponse(ok: true, output: String(Int(level * 100)))
    }
}
